src/xmerl_dsig.erl

```diff
@@ -0,0 +1,730 @@
+                            CertBin)
+    end.
+
+sign_signatures([Signature|Tail], ElementIn, PrivateKeyFun, CertBin) ->
+    PrivateKey = #'RSAPrivateKey'{} = PrivateKeyFun(),
+    logger:debug("Signature: ~p~n", [Signature]),
+    TransformAlgo =
+        xmerl_xpath:string(
+          "//*[\"Transform\"=local-name() and \"http://www.w3.org/2000/09/"
+          "xmldsig#\"=namespace-uri()]/@Algorithm", Signature),
+    [#xmlAttribute{value = SignatureMethodAlgorithm}] =
+        xmerl_xpath:string(
+          "//*[\"SignatureMethod\"=local-name() and \"http://www.w3.org/2000/"
+          "09/xmldsig#\"=namespace-uri()]/@Algorithm", Signature),
+    {HashFunction, _DigestMethod, SignatureMethodAlgorithm} =
+        signature_props(SignatureMethodAlgorithm),
+
+    [#xmlAttribute{value = "http://www.w3.org/2001/10/xml-exc-c14n#"}] =
+        xmerl_xpath:string(
+          "//*[\"CanonicalizationMethod\"=local-name() and \"http://www.w3.org/"
+          "2000/09/xmldsig#\"=namespace-uri()]/@Algorithm", Signature),
+    [#xmlAttribute{value = SignatureMethodAlgorithm}] =
+        xmerl_xpath:string(
+          "//*[\"SignatureMethod\"=local-name() and \"http://www.w3.org/2000/"
+          "09/xmldsig#\"=namespace-uri()]/@Algorithm", Signature),
+    [_C14nTx = #xmlElement{}] =
+        xmerl_xpath:string(
+          "//*[\"Transform\"=local-name() and \"http://www.w3.org/2000/09/"
+          "xmldsig#\"=namespace-uri()]", Signature),
+         %% xmerl_xpath:string(
+         %%  "ds:Signature/ds:SignedInfo/ds:Reference/ds:Transforms/ds:Transform"
+         %%  "[@Algorithm='http://www.w3.org/2001/10/xml-exc-c14n#']",
+         %%  Signature, [{namespace, DsNs}]),
+    TransformAlgo =
+        xmerl_xpath:string(
+          "//*[\"Transform\"=local-name() and \"http://www.w3.org/2000/09/"
+          "xmldsig#\"=namespace-uri()]/@Algorithm", Signature),
+
+
+    %% first we need the digest, to generate our SignedInfo element
+    [#xmlAttribute{value = RefUri}] =
+        xmerl_xpath:string(
+          "//*[\"Reference\"=local-name() and \"http://www.w3.org/2000/09/"
+          "xmldsig#\"=namespace-uri()]/@URI", Signature),
+    logger:debug("Ref: ~p~n", [RefUri]),
+    StrippedXml = strip_it(RefUri, TransformAlgo, Signature, ElementIn),
+    CanonishXml = xmerl:export_simple([StrippedXml], xmerl_xml),
+    <<"<?xml version=\"1.0\"?>", CanonXmlUtf8/binary>> =
+        unicode:characters_to_binary(CanonishXml, unicode, utf8),
+    %%CanonXml = xmerl_c14n:c14n(StrippedXml, false, InclNs),
+    %%CanonXmlUtf8 = unicode:characters_to_binary(CanonXml, unicode, utf8),
+    logger:debug("DigestInput:~n~ts~n --~n",
+                  [CanonXmlUtf8]),
+    DigestValue = crypto:hash(HashFunction, CanonXmlUtf8),
+    %% CanonXml = xmerl_c14n:c14n(ElementStrip),
+    %% DigestValue = base64:encode_to_string(
+    %%     crypto:hash(HashFunction,
+    %%                 unicode:characters_to_binary(CanonXml, unicode, utf8))),
```

Comment:
What's with these out-commented canonicalisation lines? And why are we not using `digest/2` here, which does canonicalisation using `xmerl_c14n` and calculates a digest value?

Reply:
That was probably from the org author. Gone now.